pulsar-function-go/pf/instance.go

```diff
@@ -29,68 +29,72 @@ import (
 	"github.com/apache/pulsar/pulsar-function-go/pb"
 )
 
-type GoInstance struct {
-	function  Function
+type goInstance struct {
+	function  function
 	context   *FunctionContext
 	producer  pulsar.Producer
 	consumers map[string]pulsar.Consumer
 	client    pulsar.Client
 }
 
-// NewGoInstance init GoInstance and init function context
-func NewGoInstance() *GoInstance {
-	goInstance := &GoInstance{
+// newGoInstance init goInstance and init function context
+func newGoInstance() *goInstance {
+	goInstance := &goInstance{
 		context:   NewFuncContext(),
 		consumers: make(map[string]pulsar.Consumer),
 	}
 	return goInstance
 }
 
-func (gi *GoInstance) StartFunction(function Function) {
+func (gi *goInstance) startFunction(function function) error {
 	gi.function = function
 	err := gi.setupClient()
 	if err != nil {
-		panic("setup client failed, please check.")
+		log.Errorf("setup client failed, error is:%v", err)
+		return err
 	}
 	err = gi.setupProducer()
 	if err != nil {
-		panic("setup producer failed, please check.")
+		log.Errorf("setup producer failed, error is:%v", err)
+		return err
 	}
 	channel, err := gi.setupConsumer()
 	if err != nil {
-		panic("setup consumer failed, please check.")
+		log.Errorf("setup consumer failed, error is:%v", err)
+		return err
 	}
 
 CLOSE:
 	for {
 		select {
 		case cm := <-channel:
 			msgInput := cm.Message
-			atMostOnce := gi.context.InstanceConf.FuncDetails.ProcessingGuarantees == pb.ProcessingGuarantees_ATMOST_ONCE
-			autoAck := gi.context.InstanceConf.FuncDetails.AutoAck
+			atMostOnce := gi.context.instanceConf.funcDetails.ProcessingGuarantees == pb.ProcessingGuarantees_ATMOST_ONCE
+			autoAck := gi.context.instanceConf.funcDetails.AutoAck
 			if autoAck && atMostOnce {
 				gi.ackInputMessage(msgInput)
 			}
 			output, err := gi.handlerMsg(msgInput)
 			if err != nil {
 				log.Errorf("handler message error:%v", err)
 				gi.nackInputMessage(msgInput)
-			} else {
-				gi.processResult(msgInput, output)
+				return err
```

Comment:
I think here the logic should be
if err != nil {
  if (effectivelyOnce) {log.fatal()}
  else if (autoAck && atleast_once) {
    nack()
  }
}

Reply:
sorry, i didn't understand this, if (effectivelyOnce),  we only print one line of logs, don't do anything? Or what do you mean here,  if (effectivelyOnce) need to stop the thread?

Reply:
Or what you mean here is the same as the logic of the sendasync() error handling in processresult(). 
```
If ( atleast_once) {
    nack(msgInput)
} else if (effectivelyOnce) {
    panic()
}

Reply:
in google logging library log(fatal) means process death. I assumed that would be the case with go as well. my mistake.
In any case, in case of any error while handleMessage, if effectively once, just panic. else if atleast_once, nack. That should do the trick

Reply:
thanks @srkukarni will fix it